### PR TITLE
pass dims to range error runtime function for better exception messages

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

**Comment by BBasile** [comment]
Isn't this the revival of a PR that was criticized for the added runtime cost ?

**Comment by thewilsonator** [comment]
Yes, but [we have the technology!](https://github.com/dlang/dmd/pull/6805#issuecomment-441362721)

**Comment by thewilsonator** [comment]
@WalterBright how do we turn the `c || arrayBoundsError` into blocks that can have `BCexit` set on them so they don't pollute the instruction stream?

File: src/dmd/e2ir.d [hunk 1]
```diff
@@ -5953,7 +5956,19 @@ elem *buildArrayBoundsError(IRState *irs, const ref Loc loc)
...
+    }
+    if(lwr is null)
+    {
+        lwr = el_long(TYsize_t, 0);
+    }
+    if(elength is null)
+    {
+        elength = el_long(TYsize_t, 0);
+    }
+    return el_bin(OPcall, TYvoid, eassert, el_params(elength, upr, lwr, eline, efile, null));
```
Comment (thewilsonator) [inline src/dmd/e2ir.d]: What signature does this generate for `_d_arrayboundsp`? Because the druntime PR thinks its something different.

**Comment by 12345swordy** [comment]
Is there a reason why this isn't being merge?

**Comment by thewilsonator** [comment]
Yes because https://github.com/dlang/druntime/pull/2377 fails.

**Comment by adamdruppe** [comment]
Compiling and running this test program gives sane results

```
import core.stdc.stdio;
import core.stdc.stdlib;

extern(C) void _d_arrayboundsp(char* file, uint line, size_t lower, size_t upper, size_t length) {         printf("%s:%d [%ld .. %ld] exceeds %ld\n", file, line, lower, upper, length);
        exit(0);
}

extern(C) int main() {
        int[] a;
        a[9] = 5;
        return 0;
}
```

I did not actually modify druntime (an extern(C) in user code overrides that from the library), but you see the signature there and it prints:

pain.d:11 [0 .. 9] exceeds 0

as expected in 32 and 64 bit builds. If I do runtime.initialize and set a length, that is correctly reflected too.

**Comment by WalterBright** [comment]
Adding things like this can sum up to an astonishing amount of bloat. The trouble with that is people then just turn off array bounds checking and go commando. The idea for assertions is to minimize their cost so users are encouraged to leave them in the release build.

**Comment by adamdruppe** [comment]
Then why do you pass the file and line number to them? That accounts for more bytes of machine code than this does.

**Comment by WalterBright** [comment]
file/line is the minimum.

**Comment by adamdruppe** [comment]
Why? Wouldn't be be just as good to just drop in an `int 3` instruction to drop into the debugger (or terminate the program if there is no debugger)? It'd sure take a lot fewer bytes and provide the same memory safety.

**Comment by WalterBright** [comment]
I suggest you compile some sample code and see what the result looks like.

**Comment by thewilsonator** [comment]
> I suggest you compile some sample code and see what the result looks like.

The code won't look any different yet, the rewrite of `assert`s  into 
```d
if (!cond)
    goto fail; 
/*rest of code*/ 
fail: _d_assert_p(__FILE,__LINE__);
``` 
doesn't apply, but it should be doable to apply that to the range error function so that we only pay the size cost, not the perf cost. If you could write extend that to  range error like you did for #7386 and #7392 that would be great (I tried to do it but couldn't get it to work)

**Comment by adamdruppe** [comment]
Have you actually tried it, Walter? The size difference is quite small.

I recompiled one of my ongoing programs, along with Phobos and druntime, all git master as of this morning. with and without the patch. Here's the total binary size difference:

before: 1420648

after: 1428840

difference: 8192 bytes (exactly 8 KB lol, i am guessing something got rounded up)

Let's check an array-heavy program of mine. This one is even bigger:

 2278568 before, 2294952 after... a 16 KB difference (again, interestingly, exactly, something must be being padded, I confess I am surprised to see that, but I don't actually know all the details of code generation)

BTW if we compare that against the dmd 2.082 build of the same user code, which comes in at 1185036 bytes for program 1 and 1894888 for program 2... 8/16 KB in exchange for very useful debugging tips is nothing next to 235,612/383,680 bytes - 230 KB of random who knows what. (Probably Phobos template bloat, but I don't actually know.)

If you're really concerned about executable sizes growing, get that beam removed from your eye. Then we can talk about the speck in mine.

**Review by wilzbach — APPROVED** [review]
I would love to see this merged.

**Comment by ZombineDev** [comment]
Excuse for being late for the party, but wouldn't it be more consistent to enable the behavior via `dmd -check=bounds=context`?

**Comment by adamdruppe** [comment]
On Sun, Jan 06, 2019 at 05:40:28PM -0800, Petar Kirov wrote:
> Excuse for being late for the party, but wouldn't it be more consistent to enable the behavior via `dmd -check=bounds=context`?

This has no breakage, so zero reason to complicate with a switch.

**Comment by rainers** [comment]
This seems to have triggered regression https://issues.dlang.org/show_bug.cgi?id=19758, the test `(a[b] & 0xFF) == 0` changed from
```
  movsx    eax,byte ptr [rdi]
  test     al,al
  jne      00000000000000F1
```
to
```
  movsx    esi,byte ptr [r8]
  test     dh,dh
  jne      0000000000000120
```
Note that the wrong register is used for the test. I'm not sure why the glue layer can effect register allocation, though.

**Pull request merged**